pallets/loans-ref/src/types.rs

```diff
@@ -0,0 +1,662 @@
+	/// Loan information
+	pub info: LoanInfo<T>,
+
+	/// Borrower account that created this loan
+	pub borrower: T::AccountId,
+}
+
+/// Data containing a closed loan for historical purposes.
+#[derive(Encode, Decode, Clone, TypeInfo, MaxEncodedLen)]
+#[scale_info(skip_type_params(T))]
+pub struct ClosedLoan<T: Config> {
+	/// Block when the loan was closed
+	closed_at: T::BlockNumber,
+
+	/// Loan information
+	info: LoanInfo<T>,
+}
+
+impl<T: Config> ClosedLoan<T> {
+	pub fn new(mut info: LoanInfo<T>) -> Result<Self, DispatchError> {
+		info.deactivate()?;
+
+		Ok(Self {
+			closed_at: frame_system::Pallet::<T>::current_block_number(),
+			info,
+		})
+	}
+}
+
+/// Data containing an active loan.
+#[derive(Encode, Decode, Clone, TypeInfo, MaxEncodedLen)]
+#[scale_info(skip_type_params(T))]
+pub struct ActiveLoan<T: Config> {
+	/// Id of this loan
+	loan_id: T::LoanId,
+
+	/// Loan information
+	info: LoanInfo<T>,
+
+	/// Borrower account that created this loan
+	borrower: T::AccountId,
+
+	/// Specify whether the loan has been writen off
+	written_off_status: WriteOffStatus<T::Rate>,
+
+	/// Date when the loans becomes active
+	origination_date: Moment,
+
+	/// Normalized debt used to calculate the outstanding debt.
+	normalized_debt: T::Balance,
+
+	/// Total borrowed amount of this loan
+	total_borrowed: T::Balance,
+
+	/// Total repaid amount of this loan
+	total_repaid: T::Balance,
+
+	/// When the loans's Present Value (PV) was last updated
+	last_updated: Moment,
+}
+
+impl<T: Config> ActiveLoan<T> {
+	pub fn new(loan_id: T::LoanId, info: LoanInfo<T>, borrower: T::AccountId, now: Moment) -> Self {
+		ActiveLoan {
+			loan_id,
+			info,
+			borrower,
+			written_off_status: WriteOffStatus::default(),
+			origination_date: now,
+			normalized_debt: T::Balance::zero(),
+			total_borrowed: T::Balance::zero(),
+			total_repaid: T::Balance::zero(),
+			last_updated: now,
+		}
+	}
+
+	pub fn loan_id(&self) -> T::LoanId {
+		self.loan_id
+	}
+
+	pub fn borrower(&self) -> &T::AccountId {
+		&self.borrower
+	}
+
+	pub fn maturity_date(&self) -> Moment {
+		self.info.schedule.maturity.date()
+	}
+
+	/// Returns a penalized version of the interest rate in an absolutely way.
+	/// This method first unpenalized the rate based on the current write off status before
+	/// penalize it with the input parameter.
+	/// `interest_rate_with(0)` with returns the original interest_rate without any penalization
+	fn interest_rate_with(&self, penalty: T::Rate) -> Result<T::Rate, ArithmeticError> {
+		self.info
+			.interest_rate
+			.ensure_sub(self.written_off_status.penalty)?
+			.ensure_add(penalty)
+	}
+
+	fn last_updated_debt(&self) -> Result<T::Balance, DispatchError> {
+		if self.last_updated == T::Time::now().as_secs() {
+			T::InterestAccrual::current_debt(self.info.interest_rate, self.normalized_debt)
+		} else {
+			T::InterestAccrual::previous_debt(
+				self.info.interest_rate,
+				self.normalized_debt,
+				self.last_updated,
+			)
+		}
+	}
```

Comment:
It's much more ergonomic if from the `pallet-loans` perspective I ask for the `last_updated_debt`, which will return the `debt` at the `last_updated` moment. It doesn't matter if `last_updated` is now or not.

Reply:
Tracking this in https://github.com/centrifuge/centrifuge-chain/issues/1203